### PR TITLE
[SPARK-45914][PYTHON] Support commit and abort API for Python data source write 

### PR DESCRIPTION
**Pull request by allisonwang-db**
### What changes were proposed in this pull request?

This PR introduces support for the commit and abort APIs for Python data source write. After this PR, users can customize their implementations for committing and aborting write operations.

### Why are the changes needed?

To support Python data source.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

New unit tests

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/execution/python/PythonDataSourceSuite.scala [hunk 1]
```diff
@@ -548,4 +548,94 @@ class PythonDataSourceSuite extends QueryTest with SharedSparkSession {
...
+         |                if row.id >= 10:
+         |                    raise Exception("invalid value")
+         |                file.write(json.dumps(row.asDict()) + "\\n")
+         |                cnt += 1
+         |        return SimpleCommitMessage(partition_id=partition_id, count=cnt)
+         |
+         |    def commit(self, messages) -> None:
+         |        status = dict(num_files=len(messages), count=sum(m.count for m in messages))
+         |
+         |        with open(f"{self.path}/success.json", "a") as file:
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/python/PythonDataSourceSuite.scala:588]: Should probably use `os.path.join` to make sure this test works in Windows too.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/python/PythonDataSourceSuite.scala [hunk 2]
```diff
@@ -548,4 +548,94 @@ class PythonDataSourceSuite extends QueryTest with SharedSparkSession {
...
+         |            file.write(json.dumps(status) + "\\n")
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/python/PythonDataSourceSuite.scala:589]: and add `os.path.sep`

**Comment by allisonwang-db** [comment]
cc @HyukjinKwon @cloud-fan @ueshin

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
@allisonwang-db mind rebasing? Otherwise, I think it's good to go.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**